packages/monitoring-system/data-processor/src/data-processor-factory.ts

```diff
@@ -0,0 +1,46 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+import {Task} from './task-service';
+import {DataProcessor} from './data-processors/data-processor-abstract';
+import {CloudLogsProcessor} from './data-processors/cloud-logs-data-processor';
+import {GCFProcessor} from './data-processors/cloud-functions-data-processor';
+import {CloudTasksProcessor} from './data-processors/cloud-tasks-data-processor';
+import {GitHubProcessor} from './data-processors/github-data-processor';
+
+export interface Factory {
+  getDataProcessor(task: Task): DataProcessor;
+}
+
+export class DataProcessorFactory implements Factory {
+  /**
+   * Return a relevant data processor for the given task
+   * @param task a processing task
+   * @throws if no appropriate data processor is found for task
+   */
+  public getDataProcessor(task: Task): DataProcessor {
+    switch (task) {
+      case Task.ProcessLogs:
+        return new CloudLogsProcessor();
+      case Task.ProcessGCF:
+        return new GCFProcessor();
+      case Task.ProcessTaskQueue:
+        return new CloudTasksProcessor();
+      case Task.ProcessGitHub:
+        return new GitHubProcessor();
+      default:
+        throw new Error(`Couldn't identify a data processor for task: ${task}`);
```

Comment:
If I am following correctly this error message is never logged. Is there benefit in not logging an error?

Reply:
This is basically future-proofing the code. If someone adds another value to the `Task` enum in the future and forgets to update the switch here, this error will be helpful in debugging.

Reply:
My question is more if the error logged/output anywhere. If we don't log it/output it, can we observe the error thrown? For instance the caller in `TaskService` catches the error but does nothing with the error. I

Reply:
ah ok, I see what you mean. So there was a reason I don't log the error in `TaskService` or even return the info in the rejected response - basically it's to avoid credentials leaks. I expected that the most probable errors in this application will be auth errors and so I didn't want to send the auth error back in a response or put it into the logs. 

But in this case, I can log it from within DataProcessorFactory

Reply:
The ideal way to do it would be to run the error through DLP and then log it, but baby steps